src/main/java/com/project/serverpool/service/ServerService.java

```diff
@@ -0,0 +1,138 @@
+package com.project.serverpool.service;
+
+import com.project.serverpool.database.AerospikeDatabase;
+import com.project.serverpool.domain.Client;
+import com.project.serverpool.domain.Server;
+import com.project.serverpool.model.dao.daoImpl.ClientDao;
+import com.project.serverpool.model.dao.daoImpl.ServerDao;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+
+@Service
+public class ServerService {
+
+    private final ServerDao serverDao = new ServerDao();
+    private final ClientDao clientDao = new ClientDao();
+
+    /*
+        hash(server):{
+                         "server": Server,
+                         "available capacity": 40,
+                         "clients wait to locate a recourse":List<Client>
+                     }
+    */
+    private Map<Integer, Map<String, Object>> waitingList = new HashMap<>();
+
+
+
+    public List<Server> getAllServerPools() {
+        return serverDao.getAll();
+    }
+
+    public  Server locateServer(Client client) throws Exception {
+        Server server = new Server();
+        boolean isRecourseLocatedInCreatingSate = false;
+        synchronized (server){
+            server = serverDao.findAvailableServer(client.getMemory());
+        }
+        clientDao.save(client);
+
+        //no available recourse in the current active servers
+        //check the servers in the creation state
+        if (server == null) {
+            //loop over waiting list array to check if is there available resource in the creating state servers
+            for (Map.Entry<Integer, Map<String, Object>> entry : waitingList.entrySet()) {
+                //ensure that only one thread is read/write the list
+                //to read the most recent data from the list
+                synchronized (waitingList){
```

Comment:
why are u synchronizing on objects? can u explain it ?

Reply:
to make sure that there is only one thread read/write on this object (waitingList) to ensure data consistency

Reply:
if there is a thread read/write on the waiting list, the current thread will wait that thread until finished to ensure that the data readed from the list is the latest data